### PR TITLE
Add an option to save the json output from rustc to pass to rust-analyzer

### PR DESCRIPTION
**Pull request by matts1**
@googleson78 originally wrote https://github.com/bazelbuild/rules_rust/pull/1657 in order to solve this problem. As discussed  in that PR, they're no longer working on this, so I offered to pick this up. This is the same PR as that one, but has some bugfixes and refactoring applied.

### CODE REVIEW

File: rust/private/rustc.bzl
Comment (UebelAndre) [inline rust/private/rustc.bzl:57]:
```suggestion
    doc = "Save json diagnostics from rustc",
```
Can you also be more descriptive here on what "json diagnostics" is? Or reference docs that describes it more?

File: rust/private/rustc.bzl
Comment (UebelAndre) [inline rust/private/rustc.bzl:58]:
```suggestion
    fields = {"output_diagnostics": "bool: Whether or not to output diagnostics."},
```

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Hi! Thanks for working on this!

Would you be able to show how `rust-analyzer` could consume this? My understanding is that this would allow your IDE to inline errors in the code? Similar to the [cargo check](https://doc.rust-lang.org/cargo/commands/cargo-check.html) behavior. But according to https://github.com/bazelbuild/rules_rust/issues/1633 there isn't any hook in `rust-analyzer` for this yet or there may not be? Is that correct?

**Comment by matts1** [comment]
We'd use a custom vscode extension (I've published it [here](https://marketplace.visualstudio.com/items?itemName=MattStark.bazel-rust-analyzer)) for bazel + rust (I talked to rust-analyzer folks, and they agreed that all non-cargo build systems should probably write their own rust-analyzer.

I have a prototype with everything working [here](https://source.chromium.org/chromiumos/chromiumos/codesearch/+/main:src/bazel/rules_cros/toolchains/rust/ide_support/), but the TLDR is:

1) A custom flycheck would be registered for rust-analyzer that just cats the output of 
2) Vscode extension would query a list of open files, and invoke `bazel run @rules_rust//tools/gen-rust-project -- --files <currently open files in vscode>`
3) `gen-rust-project` would convert those files to bazel targets
4) `gen-rust-project` would use bazel cquery to calculate the rustc-output files.
5) `gen-rust-project` would output a rust-project.json for the bazel targets from step 3
6) `gen-rust-project` would watch for changes to any of those files from step 4
7) When `gen-rust-project` sees a change, it concatenates all the files from step 4 together to `concatenated.rustc-output`, then writes a line to stdout, which the vscode extension then receives.
8) When the vscode extension receives the message, it invokes the vscode command `clear flychecks` and then the command `run flychecks`. We would use a custom flycheck command which just cats `concatenated.rustc-output`.

**Comment by purkhusid** [comment]
@matts1 Have you been using this patch successfully in your project?

**Comment by matts1** [comment]
Yeah, it's been working fine for me.

**Comment by Silcet** [comment]
I'm crazy interested on this one, but I can see it has stalled a bit. Can I help somehow? I have forked the repo and rebased your changes in to solve the conflicts, so I could submit my branch if that helps... :smiley:

**Comment by matts1** [comment]
@UebelAndre Could you please take a look again. CI is failing for unrelated reasons (`unsupported toolchain 'ubuntu1604-bazel-java8'`). It appears someone removed the old version of java from RBE.

You can try it out by the following:
```
cd examples/bzlmod/hello_world
bazel build //:hello_world --@rules_rust//:output_diagnostics=true --output_groups=+rustc_output,+rustc_rmeta_output
```

**Comment by purkhusid** [comment]
I think this would be awesome to get this change in, any chance that any of the maintainers could give it some time?

**Comment by matts1** [comment]
I just fixed the merge conflicts, so it should be good to go as soon as it gets approved.

**Review by matts1 — COMMENTED** [review]
@UebelAndre Your changes requested doesn't appear to be actual changes, but I can't work out how to resolve that from github's perspective. Should be fine to review.

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Sorry for the delay here. I think this seems good 😄. Just one question.

**Comment by purkhusid** [comment]
Any chance we can get this on main?

**Review by UebelAndre — APPROVED** [review]
Sorry for the delay here 😅 

I think this looks good by my final request would be updating `//docs:rust_analyzer.vm` to describe the steps mentioned here: https://github.com/bazelbuild/rules_rust/pull/1942#issuecomment-1533905077

**Comment by matts1** [comment]
@UebelAndre Should be good to go

**Review by UebelAndre — APPROVED** [review]
Thanks!

**Review by UebelAndre — APPROVED** [review]

**Pull request merged**
